feat: Redis Stream 연동을 위한 기반 코드 작성

yudonggeun opened the pull request:
## 📄 Summary
- `LazyAuctioneer` 생성
    - 경매 입찰 완료 처리를 미루고 요청을 우선적으로 대기큐에 적제하는 방식을 적용합니다.
- `RedisStreamConsumer` 생성
    - 대기큐에서 요청을 조회하는 역할을 수행하는 클래스입니다.
- `PendingMessageConsumer` 생성
   - Pending 상태인 메시지가 일정기간동안 소비되지 않는다면 재처리하는 스케줄러입니다.

![image](https://github.com/user-attachments/assets/7a84563e-e8d9-4e92-a8eb-c598141dbd18)


## 📝 Reference
[Spring data redis : stream 문서](https://docs.spring.io/spring-data/redis/reference/redis/redis-streams.html#redis.streams.send)
[누군가의 Spring boot 연동 블로그](https://kingjakeu.github.io/springboot/2022/02/10/spring-boot-redis-stream)
[Redis Stream 관련 공식 문서](https://redis.io/docs/latest/develop/data-types/streams/)


close #292

yudonggeun committed:
feat: redis stream 빈 설정

yudonggeun committed:
feat: redis stream consume 클래스 생성

- RedisStreamConsumer 소비자 클래스입니다.
    - RedisStream에서 메시지를 읽고 pending 상태로 만듭니다.

- RedisOperator는 redis의 명령을 추상화하여 다른 클래스에서 쉽게 사용할 수 있도록합니다.

yudonggeun committed:
chore: RedisStream 설정 분리

- RedisStreamConsumer에서의 streamkey, consumer group name을 application.yaml에서 관리하도록 변경합니다.

yudonggeun committed:
fix: @Value 매핑 오류 수정

yudonggeun committed:
feat: LazyAuctioneer 생성

### LazyAuctioneer
- 경매 입찰 과정을 비동기로 나중에 처리하는 작업을 실행하기 위한 클래스입니다.

### AuctionPurchaseRequestMessage 데이터 클래스
- 경매 입찰 프로세스에 필요한 최소한의 데이터를 가지고 있는 데이터 클래스입니다.

yudonggeun committed:
Merge branch 'dev' into feat/292-redis-stream

yudonggeun committed:
feat: Redis stream message ack 로직 추가

yudonggeun committed:
feat: RedisStreamConfig 설정 통합

- RedisStreamConsumer로 부터 Stream에 필요한 정보를 조회하도록 변경하였습니다.
- streamKey, consumerGroupName, consumerName이 config에 추가되었습니다.

yudonggeun committed:
feat: XPENDING, XCLAIM redis operator 구현

yudonggeun committed:
feat: 미처리 PENDING 메시지 처리 스케줄러 구현

- 1분 이상 미처리 상태인 메시지를 처리하는 스케줄러를 구현하였습니다.

yudonggeun committed:
Merge remote-tracking branch 'origin/feat/292-redis-stream' into feat/292-redis-stream

chhs2131 commented on src/main/java/com/wootecam/luckyvickyauction/global/dto/AuctionPurchaseRequestMessage.java:
```diff
@@ -0,0 +1,13 @@
+
+import java.time.LocalDateTime;
+
+public record AuctionPurchaseRequestMessage(
+        String requestId,
+        Long buyerId,
+        Long auctionId,
+        Long price,
+        Long quantity,
+        LocalDateTime requestTime
+) {
```
스트림으로 보낼 구매 요청 메세지를 이곳에 담는군요..! requestId는 UUID! 좋습니다!

chhs2131 commented on src/main/java/com/wootecam/luckyvickyauction/consumer/presentation/RedisStreamConsumer.java:
```diff
@@ -0,0 +1,75 @@
+    private StreamMessageListenerContainer<String, MapRecord<String, Object, Object>> listenerContainer;
+    private Subscription subscription;
+
+    @Override
+    public void onMessage(MapRecord<String, Object, Object> message) {
+        log.info("MessageId: {}", message.getId());
+        log.info("Stream: {}", message.getStream());
+        log.info("Body: {}", message.getValue());
+        redisOperator.acknowledge(redisStreamConfig.getConsumerGroupName(), message);
+    }
```
해당 메서드에서 `RedisStream`을 통해 전달된 메세지를 확인하고,
메시지에 맞는 로직을 호출한 뒤 정상처리되면 `redisOperator.ack`를 보내는 식으로 동작하는 것인가요?!

yudonggeun replied:
정확합니다!

chhs2131 reviewed (commented):
역시 살아숨쉬는 레디스 전자사전~~ 📀

yudonggeun committed:
fix: test 실패 오류 수정

- 테스트 환경에서만 적용되는 값만 치환해서 사용할 수 있도록 application.yaml을 application-test.yaml으로 변경
- ServiceTest, RepositoryTest 프로파일 test 적용

yudonggeun committed:
refactor: Auctioneer 메서드 파라미터 변경

yudonggeun committed:
Merge branch 'refs/heads/dev' into feat/292-redis-stream

# Conflicts:
#	src/main/java/com/wootecam/luckyvickyauction/core/auction/service/auctioneer/BasicAuctioneer.java
#	src/main/java/com/wootecam/luckyvickyauction/core/auction/service/auctioneer/LettuceLockAuctioneerProxy.java
#	src/main/java/com/wootecam/luckyvickyauction/core/auction/service/auctioneer/RedissonLockAuctioneerProxy.java

yudonggeun committed:
fix: AutioneeorProxy 제거

yudonggeun committed:
test: 테스트 코드 버그 수정

chhs2131 reviewed (approved):
🔥🔥🔥🔥🔥🔥🔥 LGTM 🔥🔥🔥🔥🔥🔥🔥
